File: python/beeai_framework/workflows/agent/agent.py
# Copyright 2025 © BeeAI a Series of LF Projects, LLC
# SPDX-License-Identifier: Apache-2.0

import random
import string
from collections.abc import Awaitable, Callable, Sequence
from typing import Any, Self, overload

from pydantic import BaseModel, InstanceOf

from beeai_framework.agents import AnyAgent
from beeai_framework.agents.experimental import RequirementAgent
from beeai_framework.agents.tool_calling.agent import ToolCallingAgent
from beeai_framework.agents.tool_calling.utils import ToolCallCheckerConfig
from beeai_framework.agents.types import (
    AgentExecutionConfig,
    AgentMeta,
)
from beeai_framework.backend.chat import ChatModel
from beeai_framework.backend.message import AnyMessage, AssistantMessage, UserMessage
from beeai_framework.context import Run
from beeai_framework.memory.base_memory import BaseMemory
from beeai_framework.memory.readonly_memory import ReadOnlyMemory
from beeai_framework.memory.unconstrained_memory import UnconstrainedMemory
from beeai_framework.tools.tool import AnyTool
from beeai_framework.utils.dicts import exclude_keys, exclude_none
from beeai_framework.utils.lists import remove_falsy
from beeai_framework.workflows.types import WorkflowRun
from beeai_framework.workflows.workflow import Workflow

AgentFactory = Callable[[ReadOnlyMemory], AnyAgent | Awaitable[AnyAgent]]
AgentWorkflowAgentType = ToolCallingAgent | RequirementAgent


class AgentWorkflowInput(BaseModel):
    input: str
    backstory: str | None = None
    expected_output: str | type[BaseModel] | None = None

    @classmethod
    def from_message(cls, message: AnyMessage) -> Self:
        return cls(input=message.text)

    def to_message(self) -> AssistantMessage:
        text = "\n\nContext:".join(remove_falsy([self.input or "", self.backstory or ""]))
        return AssistantMessage(text)


class Schema(BaseModel):
    inputs: list[InstanceOf[AgentWorkflowInput]]
    current_input: InstanceOf[AgentWorkflowInput] | None = None
    final_answer: str | None = None
    new_messages: list[InstanceOf[AnyMessage]] = []


class AgentWorkflow:
    def __init__(self, name: str = "AgentWorkflow") -> None:
        self._workflow = Workflow(name=name, schema=Schema)

    @property
    def workflow(self) -> Workflow[Schema]:
        return self._workflow

    def run(self, inputs: Sequence[AgentWorkflowInput | AnyMessage]) -> Run[WorkflowRun[Any, Any]]:
        schema = Schema(
            inputs=[
                input if isinstance(input, AgentWorkflowInput) else AgentWorkflowInput.from_message(input)
                for input in inputs
            ],
        )
        return self.workflow.run(schema)

    def del_agent(self, name: str) -> "AgentWorkflow":
        self.workflow.delete_step(name)
        return self

    @overload
    def add_agent(
        self,
        /,
        *,
        name: str | None = None,
        role: str | None = None,
        llm: ChatModel,
        instructions: str | None = None,
        tools: list[InstanceOf[AnyTool]] | None = None,
        execution: AgentExecutionConfig | None = None,
        save_intermediate_steps: bool = True,
        meta: AgentMeta | None = None,
        tool_call_checker: ToolCallCheckerConfig | bool | None = None,
        final_answer_as_tool: bool | None = None,
    ) -> "AgentWorkflow": ...
    @overload
    def add_agent(self, instance: ToolCallingAgent | RequirementAgent, /) -> "AgentWorkflow": ...
    def add_agent(
        self,
        instance: ToolCallingAgent | RequirementAgent | None = None,
        /,
        *,
        name: str | None = None,
        role: str | None = None,
        llm: ChatModel | None = None,
        instructions: str | None = None,
        tools: list[InstanceOf[AnyTool]] | None = None,
        execution: AgentExecutionConfig | None = None,
        save_intermediate_steps: bool = True,
        meta: AgentMeta | None = None,
        tool_call_checker: ToolCallCheckerConfig | bool | None = None,
        final_answer_as_tool: bool | None = None,
    ) -> "AgentWorkflow":
        if instance is None and llm is None:
            raise ValueError("Either instance or the agent configuration must be provided!")

        if not execution:
            execution = AgentExecutionConfig(
                max_retries_per_step=3,
                total_max_retries=3,
                max_iterations=20,
            )

        async def create_agent(memory: BaseMemory) -> ToolCallingAgent | RequirementAgent:
            if instance is not None:
                new_instance = await instance.clone()
                new_instance.memory = memory
                return new_instance

            return ToolCallingAgent(
                llm=llm,  # type: ignore
                tools=tools,
                memory=memory,
                save_intermediate_steps=save_intermediate_steps,
                tool_call_checker=tool_call_checker if tool_call_checker is not None else True,
                final_answer_as_tool=final_answer_as_tool if final_answer_as_tool is not None else True,
                meta=meta
                if meta
                else AgentMeta(
                    name=name or "ToolCallingAgent",
                    description=role if role else instructions if instructions else "helpful agent",
                    tools=tools or [],
                ),
                templates={
                    "system": lambda template: template.update(
                        defaults=exclude_none({"instructions": instructions, "role": role})
                    )
                },
            )

        async def step(state: Schema) -> None:
            memory = UnconstrainedMemory()
            await memory.add_many(state.new_messages)

            last_message = memory.messages[-1].text if memory.messages else ""
            run_input = state.inputs.pop(0).model_copy() if state.inputs else AgentWorkflowInput(input=last_message)
            state.current_input = run_input
            agent = await create_agent(memory.as_read_only())
            run_output = await agent.run(
                run_input.input, **exclude_keys(run_input.model_dump(), {"input"}), **execution.model_dump()
            )

            state.final_answer = run_output.message.text
            if run_input.input:
                state.new_messages.append(UserMessage(run_input.input))
            _memory = run_output.context.get("memory") if run_output.context else None
            state.new_messages.extend(_memory.messages[-2:] if _memory else [])

        self.workflow.add_step(name or f"Agent{''.join(random.choice(string.ascii_letters) for _ in range(4))}", step)
        return self

File: typescript/src/agents/requirement/agent.ts
/**
 * Copyright 2025 © BeeAI a Series of LF Projects, LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { BaseAgent } from "@/agents/base.js";
import { AnyTool } from "@/tools/base.js";
import { BaseMemory } from "@/memory/base.js";
import { AgentMeta } from "@/agents/types.js";
import { Emitter } from "@/emitter/emitter.js";
import { ChatModel } from "@/backend/chat.js";
import type { GetRunContext, MiddlewareType } from "@/context.js";
import { UnconstrainedMemory } from "@/memory/unconstrainedMemory.js";
import { Message, UserMessage } from "@/backend/message.js";
import { shallowCopy } from "@/serializer/utils.js";
import { mapObj } from "@/internals/helpers/object.js";
import { PromptTemplate } from "@/template.js";
import {
  RequirementAgentCallbacks,
  RequirementAgentExecutionConfig,
  RequirementAgentOutput,
  RequirementAgentRunInput,
  RequirementAgentRunOptions,
  RequirementAgentTemplates,
} from "./types.js";
import {
  RequirementAgentSystemPrompt,
  RequirementAgentTaskPrompt,
  RequirementAgentToolErrorPrompt,
  RequirementAgentToolNoResultPrompt,
} from "./prompts.js";
import { Requirement } from "./requirements/requirement.js";
import { RequirementAgentRunner } from "./runner.js";
import { ToolCallChecker, ToolCallCheckerConfig } from "./utils/toolCallChecker.js";
import { isDefined, isEmptyish, pickBy } from "remeda";
import { castArray } from "@/internals/helpers/array.js";
import type { RequiredExcept } from "@/internals/types.js";

export type RequirementAgentTemplateFactory<K extends keyof RequirementAgentTemplates> = (
  template: RequirementAgentTemplates[K],
) => RequirementAgentTemplates[K];

export interface RequirementAgentInput {
  llm: ChatModel;
  memory?: BaseMemory;
  tools?: AnyTool[];
  requirements?: Requirement[];
  name?: string;
  description?: string;
  role?: string;
  instructions?: string | string[];
  notes?: string | string[];
  toolCallChecker?: boolean | ToolCallCheckerConfig;
  finalAnswerAsTool?: boolean;
  saveIntermediateSteps?: boolean;
  execution?: RequirementAgentExecutionConfig;
  templates?: Partial<{
    [K in keyof RequirementAgentTemplates]:
      | RequirementAgentTemplates[K]
      | RequirementAgentTemplateFactory<K>;
  }>;
  middlewares?: MiddlewareType<RequirementAgent>[];
}

type InternalRequirementAgentInput = RequiredExcept<
  RequirementAgentInput,
  | "role"
  | "instructions"
  | "name"
  | "description"
  | "templates"
  | "notes"
  | "execution"
  | "middlewares"
>;

/**
 * RequirementAgent - A declarative AI agent with rule-based constraints
 *
 * The RequirementAgent provides predictable, controlled execution behavior across
 * different language models through rule-based constraints (Requirements).
 * Requirements can be configured as strict or flexible as necessary, adapting to
 * task requirements while ensuring consistent execution regardless of the underlying
 * model's reasoning or tool-calling capabilities.
 */
export class RequirementAgent extends BaseAgent<
  RequirementAgentRunInput,
  RequirementAgentOutput,
  RequirementAgentRunOptions
> {
  public readonly emitter = Emitter.root.child<RequirementAgentCallbacks>({
    namespace: ["agent", "requirement"],
    creator: this,
  });

  protected readonly input: InternalRequirementAgentInput;

  protected runner: new (
    ...args: ConstructorParameters<typeof RequirementAgentRunner>
  ) => RequirementAgentRunner = RequirementAgentRunner;

  constructor(input: RequirementAgentInput) {
    super();
    this.input = {
      ...input,
      memory: input.memory || new UnconstrainedMemory(),
      tools: input.tools || [],
      requirements: input.requirements || [],
      saveIntermediateSteps: input.saveIntermediateSteps ?? true,
      finalAnswerAsTool: input.finalAnswerAsTool ?? true,
      toolCallChecker: input.toolCallChecker ?? true,
    };
    if (input.middlewares) {
      this.middlewares.push(...input.middlewares);
    }
  }

  static {
    this.register();
  }

  protected async _run(
    input: RequirementAgentRunInput,
    options: RequirementAgentRunOptions = {},
    run: GetRunContext<typeof this>,
  ): Promise<RequirementAgentOutput> {
    const execution: RequirementAgentExecutionConfig = {
      maxRetriesPerStep: 3,
      totalMaxRetries: 20,
      maxIterations: 20,
      ...this.input.execution,
      ...options.execution,
    };

    const toolCallChecker = this.createToolCallChecker();

    const runner = new this.runner(
      this.input.llm,
      execution,
      this.input.tools || [],
      this.input.requirements || [],
      input.expectedOutput || null,
      toolCallChecker,
      run,
      this.input.finalAnswerAsTool,
      this.templates,
    );

    // Process input messages
    const newMessages = this.processInput(input);
    await runner.addMessages([...this.memory.messages]);
    await runner.addMessages(newMessages);

    // Run the agent
    const finalState = await runner.run();

    // Update memory
    if (this.input.saveIntermediateSteps) {
      this.memory.reset();
      await this.memory.addMany(finalState.memory.messages);
    } else {
      await this.memory.addMany(newMessages);
      // Add last tool call pair
      const messages = finalState.memory.messages;
      const lastPair = messages.slice(-2);
      await this.memory.addMany(lastPair);
    }

    if (!finalState.answer) {
      throw new Error("Agent did not produce a final answer");
    }

    return {
      result: finalState.answer,
      memory: finalState.memory,
      state: finalState,
    };
  }

  protected processInput(input: RequirementAgentRunInput): Message[] {
    if (!input.prompt) {
      return [];
    }

    const userMessage = new UserMessage(
      this.templates.task.render({
        prompt: input.prompt,
        context: input.context,
        expectedOutput: typeof input.expectedOutput === "string" ? input.expectedOutput : undefined,
      }),
    );

    return [userMessage];
  }

  protected createToolCallChecker(): ToolCallChecker {
    const config: ToolCallCheckerConfig = {};

    if (typeof this.input.toolCallChecker === "object") {
      Object.assign(config, this.input.toolCallChecker);
    }

    const checker = new ToolCallChecker(config);
    checker.enabled = this.input.toolCallChecker !== false;
    return checker;
  }

  get meta(): AgentMeta {
    const tools = this.input.tools || [];

    return {
      name: this.input.name ?? "Requirement",
      tools,
      description:
        this.input.description ??
        "The RequirementAgent is a declarative AI agent implementation that provides predictable, " +
          "controlled execution behavior across different language models through rule-based constraints.",
      ...(tools.length > 0 && {
        extraDescription: [
          `Tools that I can use to accomplish given task.`,
          ...tools.map((tool) => `Tool '${tool.name}': ${tool.description}.`),
        ].join("\n"),
      }),
    };
  }

  protected get templates(): RequirementAgentTemplates {
    const overrides = this.input.templates || {};
    const finalized = mapObj({
      system: RequirementAgentSystemPrompt,
      task: RequirementAgentTaskPrompt,
      toolError: RequirementAgentToolErrorPrompt,
      toolNoResult: RequirementAgentToolNoResultPrompt,
    } as RequirementAgentTemplates)((
      key,
      defaultTemplate: RequirementAgentTemplates[typeof key],
    ) => {
      const override = overrides[key] ?? defaultTemplate;
      if (override instanceof PromptTemplate) {
        return override;
      }
      return override(defaultTemplate) ?? defaultTemplate;
    });

    if (this.input.role || this.input.instructions || this.input.notes) {
      finalized.system.update({
        defaults: pickBy(
          {
            role: this.input.role,
            instructions: isEmptyish(this.input.instructions)
              ? undefined
              : castArray(this.input.instructions)
                  .map((v) => `- ${v}`)
                  .join("\n"),
            notes: isEmptyish(this.input.notes)
              ? undefined
              : castArray(this.input.notes)
                  .map((v) => `- ${v}`)
                  .join("\n"),
          },
          isDefined,
        ),
      });
    }

    return finalized;
  }

  createSnapshot() {
    return {
      ...super.createSnapshot(),
      input: shallowCopy(this.input),
      emitter: this.emitter,
      runnerClass: this.runner,
    };
  }

  set memory(memory: BaseMemory) {
    this.input.memory = memory;
  }

  get memory(): BaseMemory {
    return this.input.memory!;
  }
}
